ipld/kamt/src/pointer.rs

```diff
@@ -0,0 +1,297 @@
+                            }
+                        }
+                    }
+                    Ok(())
+                }
+                w if 2 <= w && w <= conf.max_array_width => {
+                    // If more child values than max width, nothing to change.
+                    let mut children_len = 0;
+                    for c in n.pointers.iter() {
+                        if let Pointer::Values(vals) = c {
+                            children_len += vals.len();
+                        } else {
+                            return Ok(());
+                        }
+                    }
+                    if children_len > conf.max_array_width {
+                        return Ok(());
+                    }
+
+                    // Collect values from child nodes to collapse.
+                    let mut child_vals: Vec<KeyValuePair<K, V>> = n
+                        .pointers
+                        .iter_mut()
+                        .filter_map(|p| {
+                            if let Pointer::Values(kvs) = p {
+                                Some(std::mem::take(kvs))
+                            } else {
+                                None
+                            }
+                        })
+                        .flatten()
+                        .collect();
+
+                    // Sorting by key, values are inserted based on the ordering of the key itself,
+                    // so when collapsed, it needs to be ensured that this order is equal.
+                    child_vals.sort_unstable_by(|a, b| {
+                        a.key().partial_cmp(b.key()).unwrap_or(Ordering::Equal)
+                    });
+
+                    // Replace link node with child values
+                    *self = Pointer::Values(child_vals);
+                    Ok(())
+                }
+                _ => Ok(()),
+            },
+            _ => unreachable!("clean is only called on dirty pointer"),
+        }
+    }
+}
+
+fn from_ipld<T: DeserializeOwned>(ipld: Ipld) -> Result<T, String> {
+    Deserialize::deserialize(ipld).map_err(|error| error.to_string())
+}
+
+fn from_ipld_map<T: DeserializeOwned>(
+    map: &mut BTreeMap<String, Ipld>,
+    key: &str,
+) -> Option<Result<T, String>> {
+    map.remove(key).map(from_ipld)
+}
+
+fn add_to_ipld_map<S: Serializer, T: Serialize>(
+    map: &mut BTreeMap<String, Ipld>,
+    key: &str,
+    value: &T,
+) -> Result<(), S::Error> {
+    let value =
+        to_ipld(value).map_err(|e| S::Error::custom(format!("cannot serialize `{key}`: {e}")))?;
+    map.insert(key.to_owned(), value);
+    Ok(())
+}
+
+/// Helper method to undo a former split.
+fn unsplit_ext(
+    conf: &Config,
+    bf: &Bitfield,
+    parent_ext: &Extension,
+    child_ext: &Extension,
+) -> Result<Extension, Error> {
+    // Figure out which bucket contains the pointer.
+    let idx = bf
+        .last_one_idx()
+        .expect("There is supposed to be exactly one pointer") as u8;
```

Comment:
I mean this method should only ever be called if there's just one pointer in one bucket. It would be a programming error to do otherwise. If there were multiple pointers, then we could not unsplit a node. 

We want to turn `A -> B -> C` into `A -> C`, but if we have `A -> B -> [C, D]`, with `B` pointing at `C` and `D`, then it's okay to leave it like that.

Reply:
So basically, this helper should not be called outside [this](https://github.com/filecoin-project/ref-fvm/pull/1094/files/20a3edacedafa5895f3de39d1bdb8f40a32e4969#diff-7b99b75f2612ac6832d2903486f74ce43fe930def46d807e728b44d698c57e17R174) branch.